annotate.el

```diff
@@ -950,18 +968,65 @@ essentially what you get from:
       (delete-overlay ov))
     (set-buffer-modified-p modified-p)))
 
-(defun annotate-create-annotation (start end &optional text)
+(defun annotate-string-empty-p (a)
+  "Is the arg an empty string or null?"
+  (or (null a)
+      (string= "" a)))
+
+(defun annotate-create-annotation (start end &optional text sample)
   "Create a new annotation for selected region."
```

Comment:
Docstring probably needs to be expanded with new functionality.

Reply:
I tried, not very happy with the results, though! :(

Reply:
Iif I'm reading the code correctly, it does

- If no text or sample are given, create a new annotation between start and end, and ask for the annotation text
- If a text is given but no sample, create a new annotation between start and end
- If both text and sample are given, and the sample matches the text between start and end, create a new annotation between start and end
- If both text and sample are given, and the sample does not match the text between start and end, search the surrounding area for a matching location, and create a new annotation there
- If both text and sample are given, and the sample does not match the text between start and end, and no matching location is nearby, silently ignore the annotation

It might be useful to split this into multiple functions (as a general rule of thumb, I try to split stuff into multiple functions if I start to feel compelled to add an "and" to the function name).

For example, we could make `text` and `sample` non-optional by changing `annotate-annotate` to grab them before calling `annotate-create-annotation`.

And then maybe add a failure condition for non-matching annotations, probably as a warning.

Reply:
Hi!
Yes i think you understand correctly the goal of the function (and i hope my code works this way :-)) 

Unfortunately i think i do not understand very well how to refactor the function, instead. :)
The two optional arguments in ``annotate-create-annotation`` are used only when the annotations are restored from the database i can not see how they can be used in `annotate-annotate`, should we change: `(annotate-create-annotation start end)` to `(annotate-create-annotation start end nil nil)`? Please help me to understand.

Anyway, good idea for the warning if annotation matching fails! I like it! :)

Bye!

Reply:
Hi!

When looking through the code, I noticed that all but one place calls `annotate-create-annotation` with four arguments. If that one place, `annotate-annotate`, would use all four arguments, we could make the last two arguments non-optional. So I propose to change `annotate-annotate` to supply the last two arguments, and to make them non-optional. This will clean up the implementation of `annotate-create-annotation` somewhat, because we won't have to include all the checks for the last two arguments any longer.

Does that explain it better?

Reply:
Hello!!

yes i think the fog in my brain is starting to going away! :-)

so, i just made the two arguments non optional and now i should move all that stuff:

``` lisp 
   (let ((annotation (or annotation-text
                          (read-from-minibuffer "Annotation: "))))
      (when (not (or (null annotation) 
                     (string= "" annotation))) ;  note this can be changed with
                                                            ; (not (string-empty-p annotation))
        (if (not (annotate-string-empty-p annotated-text))
 ```
up to `annotate-annotate`, is it?

Bye!
C.

Reply:
That's the idea!

Except of course you won't need all the conditionals, as `annotate-annotate` knows it has to ask for the annotation-string and can grab the annotated-text from the buffer.